src/scheduler_multi_app.cpp

```diff
@@ -20,13 +53,84 @@ std::vector<size_t> multi_app_sched::check_membw(const controllerT::execute_conf
 
 		// 	membw ok?
 		// 		no -> mark it
-		if (membw > 0.9) {
+		if (membw > PER_MACHINE_TH) {
 			marked_machines.push_back(c.first);
 		}
 	}
 	return marked_machines;
 }
 
+controllerT::execute_config multi_app_sched::generate_new_config(const controllerT::execute_config old_config,
+																 const std::vector<size_t> marked_machines,
+																 const std::vector<size_t> swap_candidates) {
+	// are there any swap candidates?
+	if (swap_candidates.empty()) {
+		return {};
+	}
+	assert(marked_machines.size() == swap_candidates.size());
+
+	std::vector<size_t> marked_machines_sorted = sort_machines_by_membw_util(marked_machines, true);
+
+	// determine sorted swap config
+	controllerT::execute_config new_config_sorted;
+	for (size_t idx = 0; idx < marked_machines.size(); ++idx) {
+		const size_t old_mach = marked_machines_sorted[idx];
+		const size_t new_mach = swap_candidates[idx];
+		const auto old_slot_it = std::find_if(old_config.begin(), old_config.end(),
+											  [&old_mach](auto &config_elem) { return config_elem.first == old_mach; });
+		const size_t old_slot = old_slot_it->second;
+		std::array<double, SLOTS>::iterator new_slot_it;
+		// determine 'optimal' slot on new_mach
+		// -> minimize membw_util variances of old_mach and new_mach
+		if (membw_util[old_mach][old_slot] < membw_util[old_mach][(old_slot + 1) % SLOTS]) { // TODO: what about more than 2 SLOTS?
+			new_slot_it = std::max_element(std::begin(membw_util[new_mach]), std::end(membw_util[new_mach]));
+		} else {
+			new_slot_it = std::min_element(membw_util[new_mach].begin(), membw_util[new_mach].end());
+		}
+
+		new_config_sorted.emplace_back(new_mach, std::distance(std::begin(membw_util[new_mach]), new_slot_it));
+	}
+	assert(new_config_sorted.size() == marked_machines.size());
+
+	// sort swap-config in accordance with order of marked_machines
+	controllerT::execute_config new_config;
+	for (auto &config_elem : old_config) {
+		auto machine_pos = std::find(marked_machines_sorted.begin(), marked_machines_sorted.end(), config_elem.first);
+		if (machine_pos != marked_machines_sorted.end()) {
+			new_config.emplace_back(new_config_sorted[machine_pos - marked_machines_sorted.begin()]);
+		} else {
+			new_config.emplace_back(config_elem);
+		}
+	}
+
+	return new_config;
+}
+
+std::vector<size_t> multi_app_sched::find_swap_candidates(const std::vector<size_t> &marked_machines) const {
+	// determine swap candidates
+	// -> sorted list of machine indices in accordance with membw_util
+	std::vector<size_t> swap_candidates(membw_util.size());
+	std::iota(swap_candidates.begin(), swap_candidates.end(), 0);
+	swap_candidates = sort_machines_by_membw_util(swap_candidates, false);
+	swap_candidates.resize(marked_machines.size());
```

Comment:
Can you move the resize directly before the return? Or maybe pass the required elements to sort_machines_by_membw_util and use std::partial_sort if that is enough for the algorithm.